main.cpp

```diff
@@ -0,0 +1,29 @@
+#include <vector>
+#include <string>
+#include "parser.h"
+#include <iostream>
+using namespace std;
+
+int main() {
+    vector<string> t1 {"the", "cat", "ran", "EOI"};
+    vector<string> t2 {"the", "dog", "ran", "EOI"};
+    vector<string> t3 {"the", "cat", "ran", "slowly", "EOI"};
+    vector<string> t4 {"the", "cat", "EOI"};
+    vector<string> t5 {"the", "smelly", "cat", "ran", "EOI"};
+    vector<string> t6 {"the", "lazy", "lazy", "lazy", "dog", "ran", "noisily", "EOI"};
+    vector<string> t7 {"the", "rabbit", "ran", "really fast", "EOI"};
+
+    parser tester;
+    cout << "A 1 means it is valid, a 0 means it is not" << endl;
+    cout << "Test Sentence 1: " <<  tester.check(t1) << endl;
+    cout << "Test Sentence 2: " <<  tester.check(t2) << endl;
+    cout << "Test Sentence 3: " <<  tester.check(t3) << endl;
+    cout << "Test Sentence 4: " <<  tester.check(t4) << endl;
+    cout << "Test Sentence 5: " <<  tester.check(t5) << endl;
+    cout << "Test Sentence 6: " <<  tester.check(t6) << endl;
+    cout << "Test Sentence 7: " <<  tester.check(t7) << endl;
```

Comment:
This is not testing. When the program runs, it produces a stream of 1s and 0s, but no indication of what is expected. At the least it should be something like

~~~ c++
typedef vector<string> strings;

void testOneSentence(strings sentence, int expected) {
    parser tester;
    int result = tester.check(sentence);
    cout << ((result == expected) ? "OK" : "ERROR") << ": ";
    for (int i = 0; i < sentence.size(); i++) {
        cout << sentence[i] << " ";
    }
    if (result != expected) {
        cout << " (expected " << expected << ", " << "got " << result << ")";
    }
    cout << endl;
}

int main() {
    vector<strings> testInput = {
      {"the", "dog", "ran", "EOI"},
      {"the", "cat", "ran", "EOI"},
      {"the", "cat", "ran", "slowly", "EOI"},
      {"the", "cat", "EOI"},
      {"the", "smelly", "cat", "ran", "EOI"},
      {"the", "lazy", "lazy", "lazy", "dog", "ran", "noisily", "EOI"},
      {"the", "rabbit", "ran", "really fast", "EOI"}
    };

    int expected[] = {1, 1, 1, 0, 1, 1, 0 };

    for (int i = 0; i < testInput.size(); i++) {
        testOneSentence(testInput[i], expected[i]);
    }
~~~

which gives

~~~
18∙42∙40≻ ./parser
OK: the dog ran EOI 
OK: the cat ran EOI 
OK: the cat ran slowly EOI 
OK: the cat EOI 
OK: the smelly cat ran EOI 
OK: the lazy lazy lazy dog ran noisily EOI 
OK: the rabbit ran really fast EOI 
~~~

Otherwise, you used a map to hold the state table, which is good.

I probably spent an hour in total on this,.

# -4 for the crazy tests and missing #include